[wip] bench par

BrzVlad opened the pull request.

BrzVlad commented:
build

BrzVlad commented:
builddeb

lewurm commented:
builddeb

BrzVlad commented:
builddeb

schani commented on mono/sgen/sgen-workers.c:
```diff
@@ -23,6 +23,8 @@ static int workers_num;
 static volatile gboolean forced_stop;
 static WorkerData *workers_data;
 
+static mono_mutex_t finished_lock;
```
Please document what this lock guards.

schani commented on mono/sgen/sgen-marksweep-drain-gray-stack.h:
```diff
@@ -180,15 +180,22 @@ COPY_OR_MARK_FUNCTION_NAME (GCObject **ptr, GCObject *obj, SgenGrayQueue *queue)
-			binary_protocol_pin (obj, (gpointer)SGEN_LOAD_VTABLE (obj), sgen_safe_object_get_size (obj));
+				first = FALSE;
+			else
+				sgen_los_pin_object (obj);
```
Don't you have to retry until it works?

schani commented on mono/sgen/sgen-gc.c:
```diff
@@ -1819,7 +1819,12 @@ major_copy_or_mark_from_roots (SgenGrayQueue *gc_thread_gray_queue, size_t *old_
+			if (major_collector.is_parallel)
+				sgen_workers_start_all_workers (&major_collector.major_ops_conc_par_start, NULL);
+			else
+				sgen_workers_start_all_workers (object_ops, NULL);
```
Why don't we get the correct `object_ops` in the first place?

BrzVlad replied:
At this point in the implementation, only the parallel concurrent context is properly implemented, since we don't have yet parallel copying (that is done later in the PR). Here we are doing a forced finish having as a context (normal object_ops) the non-parallel concurrent_finish context. The forced finish means that first each worker clears it's leftovers from the private gray queues, which can only be done with a context that supports parallelism.

schani commented on mono/sgen/sgen-workers.c:
```diff
@@ -433,4 +433,10 @@ sgen_workers_get_idle_func_object_ops (void)
+int
+sgen_workers_get_job_split_count (void)
+{
+	return (workers_num > 1) ? workers_num * 4 : 1;
```
Please add explanation:

If we have a single worker, splitting into multiple jobs makes no sense.  With more than one worker, we split into a larger number of jobs so that, in case the work load is uneven, a worker that finished quickly can take up more jobs than another one.

schani commented on mono/sgen/sgen-marksweep.c:
```diff
@@ -693,6 +694,56 @@ major_alloc_object (GCVTable vtable, size_t size, gboolean has_references)
 /*
+ * This can only be called by sgen workers. While this is called we assume
+ * that no other thread is accessing the block free lists. The world should
+ * be stopped and the gc thread should be waiting for workers to finish.
```
Is it necessary for this code to work that the GC thread is waiting for workers to finish?

BrzVlad replied:
As of right now yes. The gc thread cannot call this function because it lacks the tls storage for the private block free list, but it's not hard to add support. I'd rather keep it this way though, having gc thread and workers run in separate stages, so we are keeping things simple.

schani commented on mono/sgen/sgen-los.c:
```diff
@@ -644,6 +644,9 @@ sgen_los_scan_card_table (CardTableScanType scan_type, ScanCopyContext ctx, int
 			if (!sgen_los_object_is_pinned (obj->data))
 				continue;
 
+			if (!obj->cardtable_mod_union)
```
What does this have to do with the rest of this commit?

schani commented on mono/sgen/sgen-protocol.c:
```diff
@@ -332,31 +332,34 @@ static void
 		goto retry;
 
-	if (InterlockedCompareExchange (&buffer->index, index + 1 + size, index) != index)
+	if (InterlockedCompareExchange (&buffer->index, index + 1 + size + (include_worker_index ? 1 : 0), index) != index)
```
Ok, at this point we should really put `index + 1 + size + (include_worker_index ? 1 : 0` into its own variable ;-)

schani commented on mono/sgen/sgen-protocol.c:
```diff
@@ -332,31 +332,34 @@ static void
 	buffer->buffer [index++] = type;
+	/* We should never change the header format */
+	if (include_worker_index)
+		buffer->buffer [index++] = (unsigned char) worker_index;
```
I think we should call `sgen_thread_pool_is_thread_pool_thread()` only down here.

schani commented on tools/sgen/sgen-grep-binprot.c:
```diff
@@ -57,13 +59,23 @@ typedef int64_t mword;
+	if (windex) {
+		if (file_version >= 2) {
+			if (read_stream (stream, windex, 1) <= 0)
+				return SGEN_PROTOCOL_EOF;
```
See, and here we're already not re-adjusting for the `+1`...

schani commented on mono/sgen/sgen-workers.c:
```diff
@@ -210,6 +210,37 @@ workers_get_work (WorkerData *data)
+
+	/* If we're parallel, steal from other workers' private gray queues  */
+	major = sgen_get_major_collector ();
+	if (major->is_parallel) {
```
Exit early if possible, to avoid nesting.  As in

    if (!major->is_parallel)
        return FALSE;

schani commented on mono/sgen/sgen-workers.c:
```diff
@@ -210,6 +210,37 @@ workers_get_work (WorkerData *data)
+		for (i = 1; i < workers_num && !section; i++) {
+			int steal_worker = (current_worker + i) % workers_num;
+			if (state_is_working_or_enqueued (workers_data [steal_worker].state))
+				section = sgen_gray_object_steal_section (&workers_data [steal_worker].private_gray_queue);
```
I'm pretty sure there should be a `break` here.

schani commented on mono/sgen/sgen-gray.c:
```diff
@@ -180,6 +221,52 @@ sgen_gray_object_dequeue_section (SgenGrayQueue *queue)
+	} else {
+		/* We have reserved for us the last element of the queue */
+		section = queue->last;
+		section->next = NULL;
```
Shouldn't this assert that `section->next == NULL`?

schani commented on mono/sgen/sgen-gray.c:
```diff
@@ -180,6 +221,52 @@ sgen_gray_object_dequeue_section (SgenGrayQueue *queue)
+		/* We have reserved for us the last element of the queue */
+		section = queue->last;
+		section->next = NULL;
+		SGEN_ASSERT (0, section, "Why we don't have any sections to steal ?");
```
This must come before we access `section->next`.

schani replied:
Also, there is no space before a question mark!

schani commented on mono/sgen/sgen-gray.c:
```diff
@@ -146,14 +161,33 @@ sgen_gray_object_dequeue (SgenGrayQueue *queue)
+		old_num_sections = InterlockedDecrement (&queue->num_sections);
+
+		if (old_num_sections <= 0) {
+			mono_os_mutex_lock (&queue->steal_mutex);
```
Wait, so if there are more sections left we allow multiple workers to steal lock-free?

schani commented on mono/sgen/sgen-gray.c:
```diff
@@ -146,14 +161,33 @@ sgen_gray_object_dequeue (SgenGrayQueue *queue)
 #endif
 
 	if (G_UNLIKELY (queue->cursor < GRAY_FIRST_CURSOR_POSITION (queue->first))) {
```
This lock-free access to `queue->cursor` bothers me.  It's written to later down, too.  It's not clear at all to me that this is safe.

BrzVlad commented on mono/sgen/sgen-workers.c:
```diff
@@ -120,12 +124,13 @@ worker_try_finish (WorkerData *data)
-			sgen_thread_pool_job_enqueue (job);
+		if (callback) {
+			finish_callback = NULL;
+			callback ();
```
yes

BrzVlad commented on mono/sgen/sgen-workers.c:
```diff
@@ -54,6 +54,7 @@ enum {
 typedef gint32 State;
 
 static SgenObjectOperations * volatile idle_func_object_ops;
+static SgenObjectOperations *idle_func_object_ops_par, *idle_func_object_ops_nopar;
```
These ops are not changed during a worker phase. They are set only during start_all_workers, which is followed by a worker signal which I assume, is enough to make it visible for the other threads., given that the workers are coming alive from a wait.

BrzVlad commented:
@schani @kumpera I have addressed the comments. I have updated the other PR : https://github.com/mono/mono/pull/3806. Let me know if there anything else to say about this. I will fix tomorrow a small regression that the PR produced and then it is ready to go.
